### DISCUSSION TITLE
Mi Band/Amazfit OSC heart rate monitor for VRChat

### DISCUSSION THREAD

**Discussion by vard88508** [vrchat-community/osc#97] [Show and tell]
# Mi Band/Amazfit OSC heart rate monitor for VRChat
![image](https://i.imgur.com/J6bFJ7u.png)  

## What is this?
This app allows you to send OSC messages of your heart rate using your smartwatch/fitness tracker to your VRChat Avatar.  
Basically you have 3 parameters to play with:
- `Heartrate` sends float value `from -1 to 1` (0-255bpm) (Use this when you need to display bpm counter)
- `Heartrate2` sends float value `from 0 to 1` (0-255bpm) (Easier to control your animations but not enough precise over network. Do not use this to display bpm counter. Use cases: making actual sound of heartbeat, making animations which speed is equal to your bpm)
- `Heartrate3` sends int value `from 0 to 255` (0-255bpm) (Useful for those who wanna bind specific event to specific heart rate. Use case: changing your outfit on avatar to sport one when your bpm goes higher than 130)

## Supported devices
It was recently discovered that the app can support not only Mi band 4/5, but many more devices! If your smartwatch/fitness tracker supports Amazfit or Zepp apps, give it a try.
### Confirmed devices list:
- Xiaomi Mi Band 4/5/6
- Amazfit Band 5/Bip S Lite

For more information, visit the GitHub repository: https://github.com/vard88508/vrc-osc-miband-hrm

If you have any questions: ask them here, or in #avatars-osc channel in VRChat discord.

**Comment by reyayer** [comment]
My Amazfit Band 5 randomly stops sending heartbeat data to the server.

**Reply by vard88508** [reply]
Unfortunately i don't know what's the problem. Maybe bluetooth signal is too weak? Also try the other apps

**Reply by reyayer** [reply]
I don't see how it could be Bluetooth as my pc is right next to the watch and what other apps do you mean? when it stops getting data the console shows this message
![image](https://user-images.githubusercontent.com/101089240/157820980-f1f604cb-003c-47a9-a434-b499167d6014.png)

**Reply by ghost** [reply]
> My Amazfit Band 5 randomly stops sending heartbeat data to the server.

What mi band are you using?

**Reply by mknickerbocker** [reply]
Make sure you keep the window maximized, or it'll time out on you, I think the browser is putting it to sleep

**Comment by oroshikirin11** [comment]
I got it to work first for a few minutes but after some time VRC suddenly stopepd receiving data. I checked to console and found the following:
 This was at the top when the program when it started.
![grafik](https://user-images.githubusercontent.com/16406138/157772599-7b896ffd-e989-43e5-9d3d-34adce220582.png)



And this looped after ot got my heartrate 
![grafik](https://user-images.githubusercontent.com/16406138/157772577-5505cc42-ac1f-4bc4-b928-a69bb07d6bee.png)
I don't know though if these errors have anything to do with my issues.

**Reply by vard88508** [reply]
Make sure you didn't close vrc-osc-hrm.exe which sends osc messages to VRChat

**Reply by oroshikirin11** [reply]
So obvious, this fixed it! Thank you!
My problem was that I set Firefox as my default browser and it opened up there when I clicked the .exe.
My idiot head just forgot to close the console after opening there and just pasted the URL to Chrome and it worked.
From there on I bookmarked it in Chrome and it worked as long as the old console from Firefox was still open. 
So apparantly I closed up my console without noticing and just tried to restart the program over the bookmark in my Chrome ...

**Comment by oroshikirin11** [comment]
I would suggest an automated timed resync with the device. The update rate slows down veeery much after a while and only a restart and/or resync of the software/device helps. 
Best would be a tickable box as an option.

**Reply by mknickerbocker** [reply]
I found the issue with my bip3, which may be the same- 

Something with active heart rate monitoring being only on when you're "doing" a workout, like walking. If you change it to workout mode, the HR monitoring works again for me, at the once a second speed or whatever

**Comment by momo-the-monster** [comment]
Thanks for sharing this, @vard88508! It's been added to the [list](https://github.com/vrchat-community/osc/wiki/Resources#heart-rate).

**Comment by Thiagoo10** [comment]
Hello, I have a question. Is there anyway in your code to know if the person using the smartband is on an physical activity using the activity_data UUID ?

**Comment by mknickerbocker** [comment]
"No Characteristics matching UUID 00000009-0000-3512-2118-0009af100700 found in Service with UUID 0000fee1-0000-1000-8000-00805f9b34fb."

Got this error, any fix for it? Not sure what I should do

**Reply by mknickerbocker** [reply]
Using a Amazfit GTR4

**Reply by Bekosantux** [reply]
I have the same problem and it was working in Miband 6.
Using a GTR Mini

**Comment by chibeeee** [comment]
does it work with an apple watch ?

**Comment by Saitura** [comment]
So i'm having issues getting this setup. i got to the point where i connected my device to the web browser, my device is an Amazfit 5. It connects but i dont see any BPM coming from the device to the browser and dont see anything in vrchat for the osc. 
![ss (2023-06-24 at 12 20 37)](https://github.com/vrchat-community/osc/assets/137650050/e7d15a15-1e32-48a8-836a-f0bad5bd15a9)
I have no clue where to go from here if there are more steps. I didn't plan to use it on on avi, i just wanted to have it on the chatbox if it was possible.

**Reply by xxxCommando** [reply]
I have the same issue with a mi band 6

![image](https://github.com/vrchat-community/osc/assets/68127762/461e821e-41da-4ae6-be64-50005fa2b238)

Is it possible to have some info ?

**Reply by xprince89** [reply]
yes please help, im facing this issue as well. it connects but still waiting for data.

**Reply by mknickerbocker** [reply]
I found the issue with my bip3, which may be the same- 

Something with active heart rate monitoring being only on when you're "doing" a workout, like walking. If you change it to workout mode, the HR monitoring works again for me.

**Comment by DangerKiddy** [comment]
Hello! Anybody tried Xiaomi Smart Band 7 or Huawei Band 7/8?

**Reply by SpeedyGwen** [reply]
I tryied the xiaomi smart band 9 active and am stuck to the browser recognising it as a bluetooth device to conect to

**Comment by luvini** [comment]
I have a Amazfit BIP model A1608 and get this error when trying to connect:

Older device support (like Mi Band 4/5) CHECKED
![image](https://github.com/vrchat-community/osc/assets/32948452/ac5579cc-8646-4ac3-923d-43aa680c04e1)

Older device support (like Mi Band 4/5) UNCHECKED
![image](https://github.com/vrchat-community/osc/assets/32948452/6bf6478d-a846-4833-abf3-4a33fd72bafd)

In case its an issue with the UUID:
![Screenshot_20230924-003543](https://github.com/vrchat-community/osc/assets/32948452/71bc49a8-9052-42db-b72f-e64548937db7)

**Comment by mknickerbocker** [comment]
Ive been using a amazfit bip3 for months with no problems but it's suddenly connecting but not sending a heart rate, please help

**Reply by mknickerbocker** [reply]
Found the issue- something with active heart rate monitoring being only on when you're "doing" a workout, like walking. If you change it to workout mode, the HR monitoring works again.

**Comment by Messias-LF** [comment]
![vivaldi_aj4IELZAcW](https://github.com/vrchat-community/osc/assets/163913697/eff60f56-0a03-4594-8a30-85256e4e76ca)

Welp... I'm trying with Amazfit band 7 sport and I receive Connection attempt failed.

**Comment by SpeedyGwen** [comment]
is there a way to get it to work with mi band 9 ?
